src/Strategy/Python/Poetry/PyProject.hs

```diff
@@ -0,0 +1,311 @@
+      (PyProjectPoetryUrlDependencySpec ds) -> sourceUrl ds
+      (PyProjectPoetryPathDependencySpec ds) -> sourcePath ds
+
+    depVersion = case deps of
+      (PoetryTextVersion ds) -> toDependencyVersion ds
+      (PyProjectPoetryDetailedVersionDependencySpec ds) -> toDependencyVersion (poetryDependencyVersion ds)
+      (PyProjectPoetryGitDependencySpec ds) -> case asum [gitTag ds, gitRev ds, gitBranch ds] of
+        Nothing -> Nothing
+        Just version -> Just $ CEq version
+      _ -> Nothing
+
+    depEnvironment = depEnvs
+    depLocations = []
+    depTags = M.empty
+
+toDependencyVersion :: Text -> Maybe VerConstraint
+toDependencyVersion dt = case parse parseConstraintExpr "" dt of
+  Left _ -> Nothing
+  Right v -> case v of
+    CEq "*" -> Nothing
+    vc -> Just vc
+
+toDependnecy :: [DepEnvironment] -> Map Text PoetryDependency -> Map Text Dependency
+toDependnecy depEnvs = M.mapWithKey $ poetryDependencyToDependency depEnvs
+
+getDependencies :: PyProject -> [Dependency]
+getDependencies project = filter ignorePython $ map snd allDeps
+  where
+    -- pyproject typically includes python as dependency that has to be ignored
+    ignorePython = (\d -> dependencyName d /= "python")
+    allDeps = case pyprojectPoetry project of
+      Nothing -> []
+      Just pp -> M.toList (prodDeps) ++ M.toList (devDeps)
+        where
+          prodDeps = toDependnecy [EnvProduction] (dependencies pp)
+          devDeps = toDependnecy [EnvDevelopment] (devDependencies pp)
+
+type Parser = Parsec Void Text
+
+symbol :: Text -> Parser Text
+symbol = L.symbol sc
+  where
+    sc :: Parser ()
+    sc = L.space (void $ some $ char ' ') empty empty
+
+-- | Consumes whitespace ` ` or tab `\t`
+whitespaceOrTab :: Parser Text
+whitespaceOrTab = takeWhileP (Just "whitespaceOrTab") isSpaceOrTab
+  where
+    isSpaceOrTab c = c == ' ' || c == '\t'
+
+-- | Poetry constraint operators
+data PoetryConstraintOperators
+  = GreaterThanOrEqual
+  | GreaterThan
+  | LessThanOrEqual
+  | LessThan
+  | Equal
+  | NotEqual
+  | WildcardAny
+  | MajorCompatible
+  | MinorCompatible
+
+-- | Parses `VerConstraint`
+parseVerConstraint :: Parser VerConstraint
+parseVerConstraint = do
+  operator <- whitespaceOrTab *> parseConstraintOperator <* whitespaceOrTab
+  versionText <- many (noneOf [andOperatorChar, orOperatorChar] <* whitespaceOrTab)
+  case operator of
+    (Equal) -> return $ CEq $ fromCharToText versionText
+    (NotEqual) -> return $ CNot $ fromCharToText versionText
+    (GreaterThanOrEqual) -> return $ CGreaterOrEq $ fromCharToText versionText
+    (GreaterThan) -> return $ CGreater $ fromCharToText versionText
+    (LessThanOrEqual) -> return $ CLessOrEq $ fromCharToText versionText
+    (LessThan) -> return $ CLess $ fromCharToText versionText
+    (MajorCompatible) -> return $ CCompatible $ fromCharToText versionText
+    (MinorCompatible) -> return $ CCompatible $ fromCharToText versionText
+    (WildcardAny) -> return $ CEq "*"
```

Comment:
You don't need parentheses here. Parentheses serve to mark the _boundaries_ of an expression. But if we're using them in a position where there is only ever one expression (i.e. the expression is not nested), we don't need to mark the boundaries.

Reply:
Whenever you type a lot of the same code over and over again, you should think about whether a refactor is possible. You should do refactors when they expose structural symmetries in the semantics of your program and domain.

In this case, for example, you're doing a lot of `return` and `fromCharToText versionText`. Really, the main difference here is `VerConstraint` constructor you're using. So how about this?

```hs
  let v = T.pack versionText
  return $ case operator of
    Equal -> CEq v
    NotEqual -> CNot v
    GreaterThanOrEqual -> CGreaterOrEq v
    GreaterThan -> CGreater v
    LessThanOrEqual -> CLessOrEq v
    LessThan -> CLess v
    MajorCompatible -> CCompatible v
    MinorCompatible -> CCompatible v
    WildcardAny -> CEq "*"
```

Reply:
Additionally, prefer `pure` instead of `return`.  They're almost always the same thing, but `pure` can be used in places that `return` can't.  `return` is also sometimes confusing for new readers, since they conflate it with a function-terminating `return` keyword.  In haskell, that is not what `return` does at all.